File: ios/FrameTrueTracker.swift
import Foundation
import UIKit

enum FrameTrueOrientation {
    case portrait
    case portraitUpsideDown
    case landscapeLeft
    case landscapeRight
    case unknown
}

struct FrameTrueResult {
    let orientation: FrameTrueOrientation
    let portraitSeconds: TimeInterval
    let landscapeSeconds: TimeInterval
    let unknownSeconds: TimeInterval
    let confidence: Double
}

final class FrameTrueTracker {
    private var durations: [FrameTrueOrientation: TimeInterval] = [:]
    private var current: FrameTrueOrientation = .unknown
    private var lastTimestamp: TimeInterval?

    func reset() {
        durations.removeAll()
        current = .unknown
        lastTimestamp = nil
    }

    func start(_ orientation: FrameTrueOrientation,
               timestamp: TimeInterval = ProcessInfo.processInfo.systemUptime) {
        reset()
        current = orientation
        lastTimestamp = timestamp
    }

    func update(_ orientation: FrameTrueOrientation,
                timestamp: TimeInterval = ProcessInfo.processInfo.systemUptime) {
        accrue(timestamp)
        current = orientation
    }

    func finish(timestamp: TimeInterval = ProcessInfo.processInfo.systemUptime) -> FrameTrueResult {
        accrue(timestamp)
        lastTimestamp = nil
        return result()
    }

    private func accrue(_ timestamp: TimeInterval) {
        guard let last = lastTimestamp else {
            lastTimestamp = timestamp
            return
        }
        let dt = max(0, timestamp - last)
        durations[current, default: 0] += dt
        lastTimestamp = timestamp
    }

    func result() -> FrameTrueResult {
        let portrait = durations[.portrait, default: 0] + durations[.portraitUpsideDown, default: 0]
        let landscape = durations[.landscapeLeft, default: 0] + durations[.landscapeRight, default: 0]
        let unknown = durations[.unknown, default: 0]
        let classified = portrait + landscape

        guard classified > 0 else {
            return .init(orientation: .unknown,
                         portraitSeconds: portrait,
                         landscapeSeconds: landscape,
                         unknownSeconds: unknown,
                         confidence: 0)
        }

        let final: FrameTrueOrientation = landscape >= portrait ? .landscapeLeft : .portrait
        return .init(orientation: final,
                     portraitSeconds: portrait,
                     landscapeSeconds: landscape,
                     unknownSeconds: unknown,
                     confidence: max(portrait, landscape) / classified)
    }
}

extension FrameTrueOrientation {
    static func from(_ deviceOrientation: UIDeviceOrientation) -> FrameTrueOrientation {
        switch deviceOrientation {
        case .portrait: return .portrait
        case .portraitUpsideDown: return .portraitUpsideDown
        case .landscapeLeft: return .landscapeLeft
        case .landscapeRight: return .landscapeRight
        default: return .unknown
        }
    }
}

File: android/FrameTrueTracker.kt
package dev.frametrue

enum class FrameTrueOrientation {
    PORTRAIT,
    PORTRAIT_UPSIDE_DOWN,
    LANDSCAPE_LEFT,
    LANDSCAPE_RIGHT,
    UNKNOWN
}

data class FrameTrueResult(
    val orientation: FrameTrueOrientation,
    val portraitSeconds: Double,
    val landscapeSeconds: Double,
    val unknownSeconds: Double,
    val confidence: Double
)

class FrameTrueTracker {
    private val durations = mutableMapOf<FrameTrueOrientation, Double>()
    private var current = FrameTrueOrientation.UNKNOWN
    private var lastNs: Long? = null

    fun reset() {
        durations.clear()
        current = FrameTrueOrientation.UNKNOWN
        lastNs = null
    }

    fun start(orientation: FrameTrueOrientation, nowNs: Long = System.nanoTime()) {
        reset()
        current = orientation
        lastNs = nowNs
    }

    fun update(orientation: FrameTrueOrientation, nowNs: Long = System.nanoTime()) {
        accrue(nowNs)
        current = orientation
    }

    fun finish(nowNs: Long = System.nanoTime()): FrameTrueResult {
        accrue(nowNs)
        lastNs = null
        return result()
    }

    private fun accrue(nowNs: Long) {
        val last = lastNs ?: run {
            lastNs = nowNs
            return
        }
        val dt = (nowNs - last).coerceAtLeast(0L) / 1_000_000_000.0
        durations[current] = (durations[current] ?: 0.0) + dt
        lastNs = nowNs
    }

    fun result(): FrameTrueResult {
        val portrait = (durations[FrameTrueOrientation.PORTRAIT] ?: 0.0) +
            (durations[FrameTrueOrientation.PORTRAIT_UPSIDE_DOWN] ?: 0.0)
        val landscape = (durations[FrameTrueOrientation.LANDSCAPE_LEFT] ?: 0.0) +
            (durations[FrameTrueOrientation.LANDSCAPE_RIGHT] ?: 0.0)
        val unknown = durations[FrameTrueOrientation.UNKNOWN] ?: 0.0
        val classified = portrait + landscape
        if (classified <= 0.0) {
            return FrameTrueResult(FrameTrueOrientation.UNKNOWN, portrait, landscape, unknown, 0.0)
        }
        val final = if (landscape >= portrait) FrameTrueOrientation.LANDSCAPE_LEFT
                    else FrameTrueOrientation.PORTRAIT
        return FrameTrueResult(
            final,
            portrait,
            landscape,
            unknown,
            maxOf(portrait, landscape) / classified
        )
    }
}
